### PR TITLE
Marketo lead import::Field-Schema-Verify

### PR DESCRIPTION
**Pull request by yashasvibajpai**
## Description of the change

> Intersect the UI mapped fields from the fields present in Marketo, and check for non-existing fields.

## Type of change
- [ ] Bug fix (non-breaking change that fixes an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Related issues

> Fix [#1]() 

### Development

- [ ] Lint rules pass locally
- [ ] The code changed/added as part of this pull request has been covered with tests
- [ ] All tests related to the changed code pass in development

### Code review 

- [ ]  This pull request has a descriptive title and information useful to a reviewer. There may be a screenshot or screencast attached
- [ ] "Ready for review" label attached to the PR and reviewers mentioned in a comment
- [ ] Changes have been reviewed by at least one other engineer
- [ ] Issue from task tracker has a link to this pull request

### CODE REVIEW

File: v0/destinations/marketo_bulk_upload/fileUpload.js [hunk 1]
```diff
@@ -15,11 +15,40 @@ const {
...
+  const fieldArrNames = [];
+  const fieldSchemaMapping = await httpGET(
+    `https://${config.munchkinId}.mktorest.com/rest/v1/leads/describe2.json`,
+    {
+      params: {
+        access_token: await getAccessToken(config)
+      }
+    }
+  );
+  fieldArr = fieldSchemaMapping.response.data.result[0].fields;
```
Comment (ItsSudip) [inline v0/destinations/marketo_bulk_upload/fileUpload.js:32]:
```suggestion
  fieldArr = get(fieldSchemaMapping,"response.data.result[0].fields");
```

File: v0/destinations/marketo_bulk_upload/fileUpload.js [hunk 2]
```diff
@@ -15,11 +15,40 @@ const {
...
+  let fieldArr = [];
+  const fieldArrNames = [];
+  const fieldSchemaMapping = await httpGET(
+    `https://${config.munchkinId}.mktorest.com/rest/v1/leads/describe2.json`,
+    {
+      params: {
+        access_token: await getAccessToken(config)
+      }
+    }
+  );
```
Comment (ItsSudip) [inline v0/destinations/marketo_bulk_upload/fileUpload.js:31]: Check if the request is successful or not.

**Review by utsabc — CHANGES_REQUESTED** [review]

File: v0/destinations/marketo_bulk_upload/fileUpload.js [hunk 3]
```diff
@@ -29,14 +62,15 @@ const getHeaderFields = config => {
...
+const getFileData = async (inputEvents, config) => {
   const input = inputEvents;
   const messageArr = [];
   let startTime;
   let endTime;
   let requestTime;
   startTime = Date.now();
-  const headerArr = getHeaderFields(config);
+  const fieldArrNames = await fieldSchema(config);
+  const headerArr = getHeaderFields(config, fieldArrNames);
```
Comment (saikumarrs) [inline v0/destinations/marketo_bulk_upload/fileUpload.js:73]: As `fieldArrNames` is only being passed to `getHeaderFields`, you can invoke it within the same method.

File: v0/destinations/marketo_bulk_upload/fileUpload.js [hunk 4]
```diff
@@ -15,11 +15,44 @@ const {
...
+    throw new CustomError("Failed to fetch Marketo Field Schema", 400);
+  }
+  return fieldArrNames;
+};
+
+const getHeaderFields = (config, fieldArrNames) => {
   const { columnFieldsMapping } = config;
+
+  columnFieldsMapping.forEach(colField => {
+    if (fieldArrNames && !fieldArrNames.includes(colField.to)) {
```
Comment (saikumarrs) [inline v0/destinations/marketo_bulk_upload/fileUpload.js:48]: Here we are checking if columnFieldsMapping is a subset of fieldArrNames. Does the destination accept data only for a partial set of fields?

Comment (yashasvibajpai) [reply v0/destinations/marketo_bulk_upload/fileUpload.js:48]: `columnFieldsMapping` stores the fields that are configured by the user on UI. `fieldArrNames` is storing all the fields that are supported by Marketo. So we are checking if the fields given by the user are actually supported by Marketo, hence the subset.

File: v0/destinations/marketo_bulk_upload/poll.js [hunk 5]
```diff
@@ -16,19 +16,18 @@ const getPollStatus = async event => {
...
-    method: "get",
     headers: {
       "Content-Type": "application/json",
       Authorization: `Bearer ${accessToken}`
     }
   };
+  const pollUrl = `https://${munchkinId}.mktorest.com/bulk/v1/leads/batch/${event.importId}.json`;
   const startTime = Date.now();
-  const resp = await httpSend(requestOptions);
+  const resp = await httpPOST(pollUrl, requestOptions);
```
Comment (utsabc) [inline v0/destinations/marketo_bulk_upload/poll.js:26]:
```suggestion
  const pollStatus = await httpPOST(pollUrl, requestOptions);
```

File: v0/destinations/marketo_bulk_upload/poll.js [hunk 6]
```diff
@@ -16,19 +16,18 @@ const getPollStatus = async event => {
...
   const endTime = Date.now();
   const requestTime = endTime - startTime;
   if (resp.success) {
-    if (resp.response && resp.response.data.success) {
+    if (resp.response && resp.response.data.data.success) {
```
Comment (utsabc) [inline v0/destinations/marketo_bulk_upload/poll.js:30]: Does poll Status response has data inside response data? Did we check actual response from Marketo?

**Comment by utsabc** [comment]
@yashasvibajpai  there are  still test cases are failing

File: v0/destinations/marketo_bulk_upload/fileUpload.js [hunk 7]
```diff
@@ -15,11 +15,44 @@ const {
...
+    }
+  );
+  if (fieldSchemaMapping && fieldSchemaMapping.success) {
+    fieldArr = fieldSchemaMapping.response.data.result[0].fields;
+    fieldArr.forEach(field => {
+      fieldArrNames.push(field.name);
+    });
+    // const fieldNames = fieldArrNames;
+  } else {
+    throw new CustomError("Failed to fetch Marketo Field Schema", 400);
```
[the remark this answers is not in the document]
Comment (utsabc) [reply v0/destinations/marketo_bulk_upload/fileUpload.js:47]: Bumping this

[the remark this answers is not in the document]
Comment (yashasvibajpai) [reply v0/destinations/marketo_bulk_upload/fileUpload.js:47]: The error response thrown as a `CustomError` on line 45.

File: v0/destinations/marketo_bulk_upload/fileUpload.js [hunk 8]
```diff
@@ -15,11 +15,44 @@ const {
...
+  const fieldSchemaMapping = await httpGET(
+    `https://${config.munchkinId}.mktorest.com/rest/v1/leads/describe2.json`,
+    {
+      params: {
+        access_token: await getAccessToken(config)
+      }
+    }
+  );
+  if (fieldSchemaMapping && fieldSchemaMapping.success) {
+    fieldArr = fieldSchemaMapping.response.data.result[0].fields;
```
[the remark this answers is not in the document]
Comment (utsabc) [reply v0/destinations/marketo_bulk_upload/fileUpload.js:39]:
```suggestion
    fieldArr = fieldSchemaMapping.response.data.result && Array.isArray(fieldSchemaMapping.response.data.result ) ? fieldSchemaMapping.response.data.result[0].fields : [];
```

File: v0/destinations/marketo_bulk_upload/fileUpload.js [hunk 9]
```diff
@@ -304,21 +350,23 @@ const getImportID = async (input, config) => {
...
-  * {
-  "importId" : <some-id>,
-  "pollURL" : <some-url-to-poll-status>,
-}
+  {
+    "importId" : <some-id>,
+    "pollURL" : <some-url-to-poll-status>,
+  }
   */
+  const fieldSchemaNames = await fetchFieldSchema(config);
```
Comment (utsabc) [inline v0/destinations/marketo_bulk_upload/fileUpload.js:358]: Les make the access token call `await getAccessToken(config)` here once and use the access token by passing it as args

File: v0/destinations/marketo_bulk_upload/fetchJobStatus.js [hunk 10]
```diff
@@ -124,15 +123,14 @@ const getWarningJobStatus = async event => {
...
-    method: "get",
     headers: {
       "Content-Type": "application/json",
       Authorization: `Bearer ${accessToken}`
     }
   };
   const startTime = Date.now();
-  const resp = await httpSend(requestOptions);
+  const warningJobStatusUrl = `https://${munchkinId}.mktorest.com/bulk/v1/leads/batch/${importId}/warnings.json`;
+  const resp = await httpPOST(warningJobStatusUrl, requestOptions);
```
[the remark this answers is not in the document]
Comment (utsabc) [reply v0/destinations/marketo_bulk_upload/fetchJobStatus.js:133]: Bumping this should we make it httpGet?

[the remark this answers is not in the document]
Comment (yashasvibajpai) [reply v0/destinations/marketo_bulk_upload/fetchJobStatus.js]: Yes, corrected to be httpGET.

File: v0/destinations/marketo_bulk_upload/fileUpload.js (hunk 9 above)
Comment (yashasvibajpai) [reply v0/destinations/marketo_bulk_upload/fileUpload.js]: getAccessToken() called once and token gets passed around.

File: v0/destinations/marketo_bulk_upload/fetchJobStatus.js [hunk 11]
```diff
@@ -18,15 +18,14 @@ const getFailedJobStatus = async event => {
...
-    method: "get",
     headers: {
       "Content-Type": "application/json",
       Authorization: `Bearer ${accessToken}`
     }
   };
+  const failedLeadUrl = `https://${munchkinId}.mktorest.com/bulk/v1/leads/batch/${importId}/failures.json`;
   const startTime = Date.now();
-  const resp = await httpSend(requestOptions);
+  const resp = await httpPOST(failedLeadUrl, requestOptions);
```
[the remark this answers is not in the document]
Comment (yashasvibajpai) [reply v0/destinations/marketo_bulk_upload/fetchJobStatus.js]: Corrected to be httpGET like pollStatus.

File: v0/destinations/marketo_bulk_upload/fileUpload.js (hunk 3 above)
Comment (yashasvibajpai) [reply v0/destinations/marketo_bulk_upload/fileUpload.js]: Marking this as resolved as now fieldArrNames is being passed throughout.

File: v0/destinations/marketo_bulk_upload/util.js [hunk 12]
```diff
@@ -1,5 +1,5 @@
 const { CustomError } = require("../../util");
-const { httpSend } = require("../../../adapters/network");
+const { httpSend, httpGET } = require("../../../adapters/network");
```
Comment (utsabc) [inline v0/destinations/marketo_bulk_upload/util.js:2]:
```suggestion
const {  httpGET } = require("../../../adapters/network");
```

**Review by utsabc — APPROVED** [review]

**Review by saikumarrs — APPROVED** [review]

**Review by ItsSudip — APPROVED** [review]

**Pull request merged**